Fix tests broken in last regen

rossabaker opened the pull request:
The generated prefixes for ambiguous values shifted when we updated the model during Amazonka, and it broke a test.

The alternative approach would be to override prefixes on ones we're known to use.  The lesson is that they're not guaranteed to be stable as the models expand.

andreyk0 commented:
Hmm, this is annoying. I guess short term you could diff the old/new models and see if it's obvious what caused the shift, reorder JSON manually for now.
Longer term we could try to make prefixes more predictable, as you've said, or try something more dramatic (with a major version bump) like https://github.com/codetalkio/codetalk.io/blob/master/resources/code/common-json-patterns/haskell-lens/src/Main.hs#L1-L2

rossabaker commented:
I tried diffing the model to get a sense of the scope, but the EC2 JSON update was tens of thousands of lines.  Diffing the generated Haskell is also a headache, because hindent no longer works.  I replaced it with brittany, which  is unhappy with pattern synonyms in the released version.  I replaced it with a GitHub dependency on brittany, which is still unhappy with a bang syntax.

I can shift just these, but there's more work to do to retain compatibility.  And it's more grim than persistently recording prefixes: two shapes that previously didn't overlap may overlap in the new model, and prefixing is currently done at the shape level.

rossabaker commented:
aws-cfn-stacks fails to compile because it's on a pre-hackathon, Earnest-updated version of the model.  It doesn't compile on the head of develop-gen now, but the rest works without this prefix madness.  Perhaps this prefix problem isn't huge in practice, even though theoretically it could be miserable.

andreyk0 commented:
> aws-cfn-stacks fails to compile because it's on a pre-hackathon, Earnest-updated version of the model. It doesn't compile on the head of develop-gen now, but the rest works without this prefix madness. Perhaps this prefix problem isn't huge in practice, even though theoretically it could be miserable.

Oh, I'm happy to go over the errors if you like. Hopefully it's not too bad but we can decide how to go about it if it is.

fiadliel commented:
If these prefixes are the main compatibility issue, it might be worth trying to maintain the current values for the next version. I feel that for a new release, this is more important than expanding the service support (if we limited to just existing services, would it be fine?).

In the longer run, I quite dislike the current scheme. A breaking change which gives memorable identifiers would be nice.

rossabaker commented:
https://gist.github.com/rossabaker/c0dbba2bcd7794c410ce731a0a9bc388

Between upstream develop and our develop, which has seen some ad hoc updates, 11 shapes disappear and 57 change prefix.  We can see the ones that hit this PR.  Of 15688 shapes upstream, 0.43% broke.

Between upstream develop and the pinned-models branch, which represents the latest without new projects, 19 shapes disappear and 165 change prefix.  That's a 1.17% breakage rate over about 17 months.  Interestingly, the EC2 changes in this PR reverted.  The rates of change may be deceptive: if we start updating on a regular basis, we might find prefixes oscillating to and fro.  

I did not do an analysis of fields that may have been dropped or changed types.

For posterity's sake, I generated the input files with the below command from the root of the repo, and processed it with the Haskell in the gist.  The haskell takes two arguments, the old prefix log and the new.

```sh
for i in gen/model/[a-z]*; do (cd gen; export MODELS=${i#gen/}; make gen 2>&1 | awk '/^:::/' | awk -F'\t' '{ print "'${MODELS#model/#}'", "\t", $2, "\t", $3 }' ); done | tee prefix.log
```

rossabaker commented:
Covered in #20.
